### DISCUSSION TITLE
Typesafety with mocking module using storybook v8

### DISCUSSION THREAD

**Discussion by jude98** [storybookjs/storybook#28316] [Help] [answered]
### Summary

I'm working with Storybook v8 and trying to mock a hook for my stories. I've followed the official documentation on [type-safe module mocking](https://storybook.js.org/docs/writing-stories/mocking-modules), but I keep encountering the error:
`Cannot find module '#hooks/useSettings.mock' or its corresponding type declarations.ts(2307)`
It will only work on absolute path import that I have configured for the project like:
`import { useDashboardLayout } from '@hooks/useSettings.mock';`

Here is what I have done:
```javascript
// src/hooks/useSettings.mock.ts
import { fn } from '@storybook/test';
import * as actual from './useSettings';

export * from './useSettings';
export const useDashboardLayout = fn(actual.useDashboardLayout).mockName('settings::useDashboardLayout');
```

component.stories.tsx
```typescript
import type { StoryObj, Meta } from '@storybook/react';

import { useDashboardLayout } from '#hooks/useSettings.mock';

import { Dashboard } from './Dashboard';

const meta = {
  component: Dashboard,
} satisfies Meta<typeof Dashboard>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Empty: Story = {
  beforeEach: () => {
    useDashboardLayout.mockReturnValue([]);
  },
};

export const Row: Story = {
  beforeEach: () => {
    useDashboardLayout.mockReturnValue(['Hello world', 'it is great to see you']);
  },
};
```

package.json
```json
{
  "imports": {
    "#hooks/useSettings": {
      "storybook": "./src/hooks/useSettings.mock.ts",
      "default": "./src/hooks/useSettings.ts"
    },
    "#*": [
      "./*",
      "./*.ts",
      "./*.tsx"
    ]
  }
}
```

The story is working properly when I import like below in stories:
`import { useDashboardLayout } from '#hooks/useSettings';`

But its not typesafe that it gives error 
`Property 'mockReturnValue' does not exist on type '() => string[]'.ts` 
at `useDashboardLayout.mockReturnValue([]);`

### Additional information

React with vite
Storybook version 8
Storybook config:
```typescripimport { join, dirname } from 'path';

/**
 * This function is used to resolve the absolute path of a package.
 * It is needed in projects that use Yarn PnP or are set up within a monorepo.
 */
function getAbsolutePath(value) {
  return dirname(require.resolve(join(value, 'package.json')));
}

/** @type { import('@storybook/react-vite').StorybookConfig } */
const config = {
  stories: ['../src/**/*.mdx', '../src/**/*.stories.@(js|jsx|mjs|ts|tsx)'],
  addons: [
    getAbsolutePath('@storybook/addon-onboarding'),
    getAbsolutePath('@storybook/addon-links'),
    getAbsolutePath('@storybook/addon-essentials'),
    getAbsolutePath('@chromatic-com/storybook'),
    getAbsolutePath('@storybook/addon-interactions'),
    getAbsolutePath('@storybook/addon-toolbars'),
    getAbsolutePath('@chakra-ui/storybook-addon'),
  ],
  framework: {
    name: getAbsolutePath('@storybook/react-vite'),
    options: {},
  },
  features: {
    emotionAlias: false,
  },
};
export default config;
```

**Comment by JReinhold** [comment, marked as answer]
Change your `package.json` to

```json
{
  "imports": {
    "#hooks/useSettings": {
      "storybook": "./src/hooks/useSettings.mock.ts",
      "default": "./src/hooks/useSettings.ts"
    },
    "#*": [
      "./src/*",
      "./src/*.ts",
      "./scr/*.tsx"
    ]
  }
}
```

Explanation:

You import `#hooks/useSettings.mock`. The resolver will look in the import map. It doesn't find an exact match, but it finds the wildcard. However the wildcard maps to `./hooks/useSettings.mock`. that doesn't exist. With my proposed change, your wildcard will correctly map to `./src/hooks/useSettings.mock`.

Alternatively you could change your import to `#src/hooks/useSettings.mock`.

Hope that helps.

**Reply by jude98** [reply]
Thank you so much for the detailed explanation and the solution. Updating the package.json with the wildcard import mapping fixed the issue perfectly.
